File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def tree2str(self, root: Optional[TreeNode]) -> str:
        def recSearch(root: Optional[TreeNode]) -> str:
            if (root.left is None) and (root.right is None):
                return f"{root.val}"
            if root.right is None:
                return f"{root.val}({recSearch(root.left)})"
            if root.left is None:
                return f"{root.val}()({recSearch(root.right)})"
            return f"{root.val}({recSearch(root.left)})({recSearch(root.right)})"
        return recSearch(root)

File: 606-construct-string-from-binary-tree/606-construct-string-from-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    string recSearch(TreeNode* root) {
        if ((root->left == NULL) & (root->right == NULL)) {
            return std::to_string(root->val);
        }
        if (root->right == NULL) {
            return std::to_string(root->val) + "(" + this->recSearch(root->left) + ")";
        }
        if (root->left == NULL) {
            return std::to_string(root->val) + "()(" + this->recSearch(root->right) + ")";
        }
        return std::to_string(root->val) + "(" + this->recSearch(root->left) + ")(" + this->recSearch(root->right) + ")";
    }
    string tree2str(TreeNode* root) {            
        return this->recSearch(root);
    }
};
